### ISSUE TITLE
Supporting "arbitrary" properties

### ISSUE THREAD

**Issue by rkurchin** [JuliaMolSim/AtomsBase.jl#43]
This has come up a couple of times now. Examples could be:
* a charge associated with every atom (atom-wise property, scalar)
* a dipole moment associated with a system (system-level property, vector)
* ...and plenty of others.

@mfherbst, I recall a discussion of a relatively flexible/generic way to do this at some point but now can't find it, or recall the details of the proposal, but since this came up again recently over at https://github.com/FermiQC/Fermi.jl/issues/120#issuecomment-1079655811, I figured I'd file an issue here for discussion.

**Comment by mfherbst** [comment]
Atom-level properties already work (see `Atom` struct). Basically just overload `getproperty` and `hasproperty`. IMHO there is nothing additional needed for that. Same thing could be done with an identical mechanism for the system. Again no need to define an extra interface as `Base` already` provides this.

The only thing we have to do in `AtomsBase` is to agree (and document) the names of the standard keys and their expected size / type of the returned fields.

**Comment by mfherbst** [comment]
In light of what I just said, let's brainstorm a little (feel free to amend this post):

# Atom
- `magnetic_moment`: Number or vector of three floats
- `pseudopotential`: String, identifier or file name of a pseudopotential file

**Comment by rkurchin** [comment]
🤦‍♀️ that's why I couldn't find the atom-level one in an issue, because we already did it. Haha, my bad!

This approach sounds great. I suppose charge and dipole moment could be system-level for molecules? As far as atom-level properties, isotopic information might be useful. I'm not sure if we want to throw in everything we can imagine now, though, or do things more sparingly on an as-needed basis...

**Comment by mfherbst** [comment]
I think we should first collect the things someone has an actual use case for to not avoid clutter

**Comment by rashidrafeek** [comment]
A property which might be relevant would be the "Topology", which stores the bonding information of the system. In many of the [formats used for classical MD](https://chemfiles.org/chemfiles/latest/formats.html#list-of-supported-formats) calculations this information is also present in the trajectory files. 

For example, The [`Frame`](https://chemfiles.org/Chemfiles.jl/latest/reference/frame/#Chemfiles.Frame) type in Chemfiles.jl which is the analogue of an `AbstractSystem`, stores `Topology` in addition to positions, velocities and the unitcell.

**Comment by jameskermode** [comment]
Here are some examples of system level properties I've used recently:
- Total energy
- Kinetic energy
- Potential energy
- Free energy (of ion-electron system)
- Charge
- Dipole moment
- Polarisability
- Stress tensor
- Strain tensor

I think it will be very hard to anticipate what users would like to store  - for example I'd often want to store multiple energies coming from DFT and a variety of more approximate models - so I'd advocate for making it as flexible as possible, but I agree there's nothing wrong with identifying and documenting some standard properties.

**Comment by cortner** [comment]
See my implementation of particle `State`s in `ACE.jl` - that's exactly what it provides. It could potentially move here, or it could inspire a new and better implementation here.

**Comment by tjjarvinen** [comment]
This is in my opinnion a good suggestion and I can see several cases where it would be useful.

In addition to this, it might be a good idea to change the type of `atomic_symbol` to a parameter. Because that would allow making the `Atom` type a `bitstype`, when using e.g. `SVector{2,Char}`. But this could have some other issues, so I am a bit unsure of it.

**Comment by cortner** [comment]
For exactly that reason we used a wrapped integer in JuLIP but allowed easy translation to a Symbol. With a bit of work the display could even show the symbol instead of the atomic number.

**Comment by lmiq** [comment]
> the type of `atomic_symbol` to a parameter.

I would follow what Cortner suggested, just store the atomic number and make the symbol, or string atomic name, something to be get on the flight.

**Comment by jameskermode** [comment]
The only issue with the last suggestion is that it requires a one to one mapping between species and atomic number, which is maybe ok for quantum mechanics but not for classical MD where there is often more than one type of each atom (even in QM there could be different pseudo potentials)

**Comment by lmiq** [comment]
That right, one more reason to allow a typed extra data field. Then the atomic type information can be stored there in any format needed, and there are many formats possible (Ints, or strings, etc).

**Comment by cortner** [comment]
Then I'll promote again my implementation in ACE.jl of wrapped NamedTuples. 

https://github.com/ACEsuit/ACE.jl/blob/main/src/states.jl

https://github.com/ACEsuit/ACE.jl/blob/main/test/test_states.jl

**Comment by jameskermode** [comment]
I did this in ExtXYZ.jl already. Could be moved here. 

https://github.com/libAtoms/ExtXYZ.jl/blob/master/src/atoms.jl

**Comment by lmiq** [comment]
The issue I see with that approach is that it does not enforce *anything*. It is almost as saying that the interface is just a commitment to the implementation of some functions. IMO the interface is useful if at all agree with at least some properties to be always present.

(I would say positions and atomic number - velocities maybe, but mass, symbol, name, are either redundant with the atomic number or have to be extended arbitrarily for new types).

But if all agree that the interface is only related to the implementation of a subtype of `AbstractAtom` and some getter and setter functions, there isn't need for this discussion, each package could implement the structure at will. In that case, the base interface should not depend on any type parameter or internal field of the `Atom` struct.

**Comment by cortner** [comment]
I thought we are talking about a prototype implementation. Indeed I don't think the interface can depend on any type parameters at all. Just on getter functions.

**Comment by cortner** [comment]
I guess if you want an abstract CHARMM then just implement a new function isCHARMM which defaults to false?

**Comment by lmiq** [comment]
But what interoperability could we get from that interface, at all? That's what I miss.

Why then `Atom` here is exported at all?

**Comment by cortner** [comment]
> But what interoperability could we get from that interface, at all?

If the getters are defined and documented in a public interface, that's all you need for interoperability I think? Is there something concrete that you can't do? 

> Why then Atom here is exported at all?

I think it is intended as a minimal prototype. 

I wouldn't be against helping implement another publicly share concrete type say `FlexibleAtom` or something similar that has more functionality. The standard functionality could be the same as for `Atom` and hence guaranteed, but if could have arbitrary additional typed field. E.g. in a wrapped NamedTuple, which would also give you the type information you want.

But I wouldn't necessarily want to *rely* on that, and rather have getters that extract the information.

**Comment by lmiq** [comment]
> I think it is intended as a minimal prototype.

Ok, so from this point of view the `Atom` struct is a minimal prototype. Then, a few questions:

Is the `Atom` constructor part of the mandatory interface? The [docs](https://docs.juliahub.com/AtomsBase/iMMMR/0.2.2/atomicsystems/#Atoms-and-FlexibleSystem) seem to suggest that. 

If the `Atom` struct is not part of the expected interface, then IMO it should not be exported. And should be better provided as an example. In any case, the present discussion is then of course unnecessary, as arbitrary properties are already accepted by any implementation of a struct for which the [list of getters](https://docs.juliahub.com/AtomsBase/iMMMR/0.2.2/apireference/) is defined. Is that the correct way to interpret this? The use or not of a symbol to represent the atomic name is also irrelevant as is only the choice for the prototype exemplified. 

In this case, wouldn't be sensible to have an `AbstractAtom` type, to at least carry the information that one expects that an atomic type defined in an independent package is a subtype of `AtomsBase.AbstractAtom` and, as such, is expected to satisfy the interface?

**Comment by cortner** [comment]
What you suggest makes sense in Java but not in Julia. Abstract super types are not needed to share methods. At the moment it is not clear what would be shared. Maybe in the future this will change.

**Comment by lmiq** [comment]
What I can imagine is something like:

```julia
julia> abstract type AbstractAtom end

julia> charge(::AbstractAtom) = "not implemented"
charge (generic function with 1 method)

julia> atomic_number(::AbstractAtom) = "not implemented"
atomic_number (generic function with 1 method)

julia> struct Atom <: AbstractAtom
           atomic_number::Int
           charge::Float64
       end

julia> a = Atom(1,1.0)
Atom(1, 1.0)

julia> charge(a)
"not implemented"
```

such that methods not implemented (but possibly expected) can be documented and provide sensible responses.

**Comment by cortner** [comment]
Why do you need this? The compiler will tell you that it isn't implemented. 

But even if you insist you can just use ::Any

**Comment by lmiq** [comment]
Well, it is a way to inform what the interface expects. And methods implemented across packages can be type annotated on the abstract type to inform that the function in case is expected to follow the interface.

I'm fine not using, but then I don't see even why a package would have to depend on AtomsBase at all.

And, back to the topic, I don't see the point of this issue in particular. The fact that this issue is open makes me believe that it is not clear to everyone what to expect from the interface.

**Comment by cortner** [comment]
We need to depend on AtomsBase to share the functions that we use to access information stored in Atoms or Structures ie systems of particles.

The interface is defined via documentation.

**Comment by lmiq** [comment]
Do people here agree, at least, that this issue of *arbitrary properties* is not really fundamental, at least? Meaning, that if one implements these 6 getters:

**Comment by cortner** [comment]
- Maybe the documentation can be improved. 

- `Atoms` is not part of the expected interface as far as I understand, but an implementation detail? I could be wrong.

**Comment by cortner** [comment]
I've said this many times in many discussions and I'll say it again : in the absence of multiple inheritence (but to some extent even *with*) any sort of `Abstract`something or other  limits you severely. Abstract supertypes are only needed to share methods, but it is *not* needed to document an interface.

**Comment by cortner** [comment]
I'm all for dropping AbstractSystem :)

**Comment by rkurchin** [comment]
Yeah, I've kind of realized that in Julia, it only makes sense in somewhat narrow cases, and usually where the taxonomy is very unambiguous – e.g. types of numbers.
